pkg/token/serviceprincipaltoken.go

```diff
@@ -59,3 +109,109 @@ func (p *servicePrincipalToken) Token() (adal.Token, error) {
 	}
 	return spt.Token(), nil
 }
+
+func isPublicKeyEqual(key1, key2 *rsa.PublicKey) bool {
+	if key1.N == nil || key2.N == nil {
+		return false
+	}
+	return key1.E == key2.E && key1.N.Cmp(key2.N) == 0
+}
+
+func splitPEMBlock(pemBlock []byte) (certPEM []byte, keyPEM []byte) {
+	for {
+		var derBlock *pem.Block
+		derBlock, pemBlock = pem.Decode(pemBlock)
+		if derBlock == nil {
+			break
+		}
+		if derBlock.Type == certificate {
+			certPEM = append(certPEM, pem.EncodeToMemory(derBlock)...)
+		} else if derBlock.Type == privateKey {
+			keyPEM = append(keyPEM, pem.EncodeToMemory(derBlock)...)
+		}
+	}
+
+	return certPEM, keyPEM
+}
+
+func parseRsaPrivateKey(privateKeyPEM []byte) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode(privateKeyPEM)
+	if block == nil {
+		return nil, fmt.Errorf("Failed to decode a pem block from private key")
+	}
+
+	privatePkcs1Key, errPkcs1 := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if errPkcs1 == nil {
+		return privatePkcs1Key, nil
+	}
+
+	privatePkcs8Key, errPkcs8 := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if errPkcs8 == nil {
+		privatePkcs8RsaKey, ok := privatePkcs8Key.(*rsa.PrivateKey)
+		if !ok {
+			return nil, fmt.Errorf("pkcs8 contained non-RSA key. Expected RSA key")
+		}
+		return privatePkcs8RsaKey, nil
+	}
+
+	return nil, fmt.Errorf("failed to parse private key as Pkcs#1 or Pkcs#8. (%s). (%s)", errPkcs1, errPkcs8)
+}
+
+func parseKeyPairFromPEMBlock(pemBlock []byte) (*x509.Certificate, *rsa.PrivateKey, error) {
+	certPEM, keyPEM := splitPEMBlock(pemBlock)
+
+	privateKey, err := parseRsaPrivateKey(keyPEM)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Mooncake certificate signed by external provider is a certificate chain, we need to get the non-CA certificate,
```

Comment:
can we omit the Azure environment specifics?

Reply:
done